### PR TITLE
v4.2

### PR DESCRIPTION
**Pull request by ArtificialErmine**
Adds compatibility with the score aggregation functionality of version 4.2.
Also improves filtering based on platforms, as well as minor stylistic fixes in various places.

### CODE REVIEW

File: mitreattack/navlayers/exporters/svg_objects.py [hunk 1]
```diff
@@ -422,8 +425,15 @@ def _com_color(self, technique, tactic, colors=[]):
...
         c = 'FFFFFF'
-        if technique.score:
-            c = self.grade.compute_color(technique.score)[1:]
+        tscore = None
+        if technique.score is not None:
+            tscore = technique.score
+        try:
+            tscore = technique.aggregateScore
+        except AttributeError:
+            pass
```
Comment (isaisabel) [inline mitreattack/navlayers/exporters/svg_objects.py:434]: Surely there must be a better way to check if the technique has an aggregate score than try-catch? If this is the best way to do it it's acceptable, it just seems like using try/catch for branching logic is bad practice.

**Review by isaisabel — CHANGES_REQUESTED** [review]
I have yet to test the code itself, but from reading it through it seems like there are a few edge cases around aggregate scores and the `countUnscored` option. I will test the code now, but figured I could submit this review first so you can get a head start on making the changes.

File: mitreattack/navlayers/core/technique.py [hunk 2]
```diff
@@ -149,6 +164,8 @@ def _loader(self, data):
                 self.metadata = data[entry]
             elif entry == 'showSubtechniques':
                 self.showSubtechniques = data[entry]
+            elif entry == 'aggregateScore':
+                self.aggregateScore = data[entry]
```
[the remark this answers is not in the document]
Comment (ArtificialErmine) [reply mitreattack/navlayers/core/technique.py:168]: Some of your test scripts had it manually set, so I left in the ability to import it. If it is, it will still always be calculated on the fly for exports. Still want me to change it?

File: mitreattack/navlayers/exporters/svg_objects.py (hunk 1 above)
Comment (ArtificialErmine) [reply mitreattack/navlayers/exporters/svg_objects.py:434]: I'll switch it to a slightly different approach.

**Comment by isaisabel** [comment]
Another thing I didn't notice -- we need to add a changelog to this repo (and document this release, `v1.1`, within said changelog). We can use the same format as on the `attack-scripts` repo.

The general format can be described in this comment to be prepended to the changelog itself:
```
<!--    CHANGELOG FORMAT                                                -->
<!--    Completed Entry template:                                       -->
<!--                                                                    -->
<!--    # Date in DD MMM YYYY format                                    -->
<!--    ## mitreattack-python version ##.##.##                          -->
<!--    ### New Features                                                -->
<!--    ### Improvements                                                -->
<!--    ### Fixes                                                       -->
<!--    VERSION NUMBERING                                               -->
<!--                                                                    -->
<!--    website versions are set up in a major.minor.patch format:      -->
<!--    MAJOR updates are when we release major new features or         -->
<!--          pages                                                     -->
<!--    MINOR updates are when we improve a small number of             -->
<!--          existing features                                         -->
<!--    PATCH updates are when a bugfix is made without the             -->
<!--          addition of notable features. When PATCH is 0 it can      -->
<!--          be omitted                                                -->
```

File: mitreattack/navlayers/core/layout.py [hunk 3]
```diff
@@ -12,6 +21,50 @@ def __init__(self):
...
+            data = data_block
+
+            if self.__aggregateFunction == Aggregates.average:
+                return sum(data)/len(data)
+            elif self.__aggregateFunction == Aggregates.min:
+                return min(data)
+            elif self.__aggregateFunction == Aggregates.max:
+                return max(data)
+            elif self.__aggregateFunction == Aggregates.sum:
+                return sum(data)
```
[the remark this answers is not in the document]
Comment (ArtificialErmine) [reply mitreattack/navlayers/core/layout.py:67]: Gotcha, I definitely misunderstood how this was supposed to work. I'll patch it appropriately.

File: mitreattack/navlayers/core/technique.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (isaisabel) [reply mitreattack/navlayers/core/technique.py:168]: Yeah I think some of the test layers were generated on an earlier iteration of the feature where they were included in layers. It should not be included as of the final Navigator version for this release.

File: mitreattack/navlayers/core/layout.py [hunk 4]
```diff
@@ -12,6 +21,50 @@ def __init__(self):
...
+        """
+            Compute the aggregate score for a technique and any subtechniques
+            :param technique: The chosen technique object
+            :param subtechniques: Any relevant subtechnique objects
+            :return: Integer representing aggregate score (or None if not computable)
+        """
+        scores = []
+        if self.showAggregateScores:
+            if technique.score is not None:
+                scores = [technique.score]
```
[the remark this answers is not in the document]
Comment (ArtificialErmine) [reply mitreattack/navlayers/core/layout.py:38]: Yeah, this will be addressed by the changes for the next comment.

File: mitreattack/navlayers/core/technique.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ArtificialErmine) [reply mitreattack/navlayers/core/technique.py:168]: Gotcha, I'll remove it.

**Comment by isaisabel** [comment]
Test script and layers (drop in root of repository and run with `python3 test.py`). Disregard the descriptions on those layers, I'm reusing them from testing on the Navigator.
[tests.zip](https://github.com/mitre-attack/mitreattack-python/files/6385302/tests.zip)

A few bugs can be observed from this (excepting the min/max bug noted already):
- example2: on the excel spreadsheet, T1548.002 has no color, but it should be red (since it has an explicit score of 0). This is correctly reflected in the svg. T1548 also may have the wrong color, it's hard to tell but I think it should be slightly more yellow (compare to the SVG, which is correct).
- example3: on the excel spreadsheet, T1548 should have a yellow color since the aggregate score should be `(100+100+100+100-100)/5 = 300/5 = 60`, but instead it has green which is the gradient of 100. This is correctly reflected in the svg.

**Comment by isaisabel** [comment]
There seems to be a bug with the average function, or maybe the order of precedence for cell coloring. In the following example layer, the technique should be orange, but is green. 

This is probably because the technique's score is used instead of the aggregate score when computing the gradient in this case. If aggregate scores are enabled, they should be displayed in the place of the technique's score-color.

Intended display:
Screen Shot 2021-04-27 at 3 17 39 PM
Actual display:
Screen Shot 2021-04-27 at 3 19 10 PM
Screen Shot 2021-04-27 at 3 19 19 PM
Layer:
```
{
    "name": "example2",
    "description": "",
    "versions": {
        "attack": "8",
        "navigator": "4.2",
        "layer": "4.1"
    },
    "selectTechniquesAcrossTactics": false,
    "domain": "enterprise-attack",
    "layout": {
        "layout": "side",
        "showID": false,
        "showName": true,
        "showAggregateScores": true,
        "aggregateFunction": "average"
    },
    "techniques": [
        {
            "techniqueID": "T1548",
            "tactic": "privilege-escalation",
            "showSubtechniques": true,
            "score": 100
        },
        {
            "techniqueID": "T1548.001",
            "tactic": "privilege-escalation",
            "score": 0
        },
        {
            "techniqueID": "T1548.002",
            "tactic": "privilege-escalation",
            "score": 0
        },
        {
            "techniqueID": "T1548.003",
            "tactic": "privilege-escalation",
            "score": 0
        },
        {
            "techniqueID": "T1548.004",
            "tactic": "privilege-escalation",
            "score": 0
        }
    ]
}
```

**Review by isaisabel — APPROVED** [review]
Seems to have resolved all the outstanding bugs.
